lib/Widgets/SeekBar.vala

```diff
@@ -16,34 +16,232 @@
+
+        /* signals */
+        scale.button_press_event.connect ((event) => {
+            is_grabbing = true;
+            scale_button_press (event);
+            return false;
+        });
+
+        scale.button_release_event.connect ((event) => {
+            is_grabbing = false;
+            set_progress (scale.get_value ());
+            scale_button_release (event);
+            return false;
+        });
 
-        var time_label = new Gtk.Label ("0:00");
+        scale.enter_notify_event.connect ((event) => {
+            is_hovering = true;
+            scale_hover (event);
+            return false;
+        });
+
+        scale.leave_notify_event.connect ((event) => {
+            is_hovering = false;
+            scale_leave (event);
+            return false;
+        });
+
+        scale.motion_notify_event.connect ((event) => {
+            set_progress (scale.get_value ());
+            scale_motion (event);
+            return false;
+        });
+
+        scale.change_value.connect ((scroll, new_value) => {
+            scroll_action (scroll, new_value);
+            return false;
+        });
+
+        scale.size_allocate.connect ((alloc_rect) => {
+            //scale_size_allocate (alloc_rect);
+        });
 
-        column_spacing = 12;
         add (progression_label);
         add (scale);
         add (time_label);
     }
 
+    /*
+     * Sets the progress of the SeekBar, with a value between 0.0 and 1.0.
+     * */
+    public void set_progress (double progress) {
+        if (progress < 0.0) {
+            warning ("Progress value less than zero, progress set to 0.0");
+            progress = 0.0;
+        }
+        else if (progress > 1.0) {
+            warning ("Progress value greater than zero, progress set to 1.0");
+            progress = 1.0;
+        }
+
+        this.playback_progress = progress;
+        scale.set_value (playback_progress);
+        progression_label.label = seconds_to_time ((int) (playback_progress*playback_duration));
+    }
+
+    /*
+     * Sets the duration of the SeekBar with a value greater than 0.0.
+     * */
+    private void set_duration (double duration) {
+        if (duration < 0.0) {
+            warning ("Duration value less than zero, duration set to 0.0");
+            duration = 0.0;
+        }
+
+        this.playback_duration = duration;
+        time_label.label = seconds_to_time ((int) duration);
+    }
+
+    public override void get_preferred_width (out int minimum_width, out int natural_width) {
+        base.get_preferred_width (out minimum_width, out natural_width);
+
+        if (parent.get_window () == null)
+            return;
+        var width = parent.get_window ().get_width ();
+        if (width > 0 && width >= minimum_width) {
+            natural_width = width;
+        }
+    }
+
+
+    /*
+     * Converts seconds into the ISO 8601 standard date format for minutes (e.g. 100s to 01:40)
+     */
+    public string seconds_to_time (int seconds) {
```

Comment:
This function doesn't use anything from the instance, hence should be `static`. I'm also not sure about the place that it's in. Perhaps that should be moved into `Granite.DateTime`?